spec/mempool/v0/quint/mempoolv0.qnt

```diff
@@ -0,0 +1,524 @@
+    action ReceiveRecheckTxResponse(node: NodeId): bool = all {
+        node.setStep("ReceiveRecheckTxResponse", Map("node" -> node)),
+        require(node.ABCI::recheckTxRequests().nonEmpty()),
+        nondet request = oneOf(node.ABCI::recheckTxRequests())
+        val response = node.ABCI::responseFor(request)
+        all {
+            node.ABCI::removeRequest(request),
+            if (response.err != noError) all {
+                // Tx became invalidated due to newly committed block.
+                node.removeFromMempool(Set(request.tx)),
+                node.removeFromCache(request.tx),
+            } else all {
+                mempool' = mempool,
+                cache' = cache,
+            }
+        },
+        node.setError(noError),
+        mempoolHeight' = mempoolHeight,
+        History::unchanged,
+        Network::unchanged,
+        Chain::unchanged,
+    }
+    //--------------------------------------------------------------------------
+    // RPC initiated actions
+    //--------------------------------------------------------------------------
+    /* Receive a specific transaction from a client via RPC. */
+    // [Environment.BroadcastTxAsync]: https://github.com/CometBFT/cometbft/blob/111d252d75a4839341ff461d4e0cf152ca2cc13d/rpc/core/mempool.go#L22
+    //-- __subaction__: this action is not part of the next-state relation
+    action CheckTxRPC(node, tx) = all {
+        node.setStep("ReceiveTxViaRPC", Map("tx" -> tx)),
+        node.CheckTx(tx, noNode),
+        History::recordSubmittedTx(node, tx),
+        Network::unchanged,
+        Chain::unchanged,
+    }
+
+    /* Receive some transaction from a client via RPC. */
+    action ReceiveTxViaRPC(node) = 
+        nondet tx = oneOf(Txs)
+        all {
+            // We don't want to repeat this step on the same transaction as the last time.
+            require(_step.name == "ReceiveTxViaRPC" and _step.node == node implies _step.args.get("tx") != tx),
+            node.CheckTxRPC(tx),
+        }
+
+    /* Receive some transaction from a client via RPC: only a transaction that
+     * was not sent before. */
+    action ReceiveNonRepeatedTxViaRPC(node) =
+        val availableTxs = Txs.exclude(History::txsReceivedViaRPC.mapValues().flatten())
+        all {
+            require(availableTxs.nonEmpty()),
+            nondet tx = oneOf(Txs.exclude(History::txsReceivedViaRPC.mapValues().flatten()))
+            node.CheckTxRPC(tx)
+        }
+
+    //--------------------------------------------------------------------------
+    // Consensus actions
+    //--------------------------------------------------------------------------
+    // Consensus reactor's BlockExecutor updates the mempool with the latest
+    // block it created and rechecks outstanding txs.
+    
+    // This node is not the proposer: take the block from the chain.
+    action BlockExecutor_MempoolUpdate_nonProposer(node) = all {
```

Comment:
Right. I will add the following to the update actions, to make sure all ABCI requests are processed before updating.
```
        require(node.ABCI::checkTxRequests().isEmpty()),
        require(node.ABCI::recheckTxRequests().isEmpty()),
```
This won't make the flush-commit-update steps atomic, as when they are hold by the lock, but it will work.